src/mod_presence.erl

```diff
@@ -0,0 +1,611 @@
+    {ok, Acc}.
+
+-spec foreign_event(Acc, Params, Extra) -> Result when
+      Acc :: mongoose_acc:t(),
+      Params :: mongoose_c2s:hook_params(),
+      Extra :: gen_hook:extra(),
+      Result :: gen_hook:hook_fn_ret(mongoose_acc:t()).
+foreign_event(Acc, #{c2s_data := StateData,
+                     event_type := info,
+                     event_content := {broadcast, {item, IJID, ISubscription}}}, _Extra) ->
+    case get_mod_state(StateData) of
+        {error, not_found} ->
+            {ok, Acc};
+        Handler ->
+            {stop, handle_subscription_change(Acc, StateData, IJID, ISubscription, Handler)}
+    end;
+foreign_event(Acc, _Params, _Extra) ->
+    {ok, Acc}.
+
+-spec handle_subscription_change(mongoose_acc:t(), mongoose_c2s:c2s_data(), term(), term(), presences()) ->
+    mongoose_acc:t().
+handle_subscription_change(Acc, StateData, IJID, ISubscription, Handler) ->
+    To = jid:make(IJID),
+    IsSubscribedToMe = (ISubscription == both) or (ISubscription == from),
+    AmISubscribedTo = (ISubscription == both) or (ISubscription == to),
+    WasSubscribedToMe = gb_sets:is_element(To, Handler#presences.pres_f),
+    FSet = case IsSubscribedToMe of
+               true ->
+                   gb_sets:add_element(To, Handler#presences.pres_f);
+               false ->
+                   gb_sets:del_element(To, Handler#presences.pres_f)
+           end,
+    TSet = case AmISubscribedTo of
+               true ->
+                   gb_sets:add_element(To, Handler#presences.pres_t);
+               false ->
+                   gb_sets:del_element(To, Handler#presences.pres_t)
+           end,
+    case Handler#presences.pres_last of
+        undefined ->
+            NewHandler = Handler#presences{pres_f = FSet, pres_t = TSet},
+            mongoose_c2s_acc:to_acc(Acc, state_mod, {?MODULE, NewHandler});
+        P ->
+            ?LOG_DEBUG(#{what => roster_changed, roster_jid => To,
+                         acc => Acc, c2s_state => StateData}),
+            From = mongoose_c2s:get_jid(StateData),
+            IsntInvisible = not Handler#presences.pres_invis,
+            ImAvailableTo = gb_sets:is_element(To, Handler#presences.pres_a),
+            ImInvisibleTo = gb_sets:is_element(To, Handler#presences.pres_i),
+            BecomeAvailable = IsntInvisible and IsSubscribedToMe and not WasSubscribedToMe,
+            BecomeUnavailable = not IsSubscribedToMe and WasSubscribedToMe
+                                and (ImAvailableTo or ImInvisibleTo),
+            case {BecomeAvailable, BecomeUnavailable} of
+                {true, _} ->
+                    ?LOG_DEBUG(#{what => become_available_to, roster_jid => To,
+                                 acc => Acc, c2s_state => StateData}),
+                    ejabberd_router:route(From, To, Acc, P),
+                    A = gb_sets:add_element(To, Handler#presences.pres_a),
+                    NewHandler = Handler#presences{pres_a = A, pres_f = FSet, pres_t = TSet},
```

Comment:
I think it's better to call it `Presences` - I think we agreed to move away from the name `Handler`.